pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1431,6 +1428,15 @@ public boolean hasMessageAvailable() throws PulsarClientException {
         return booleanFuture;
     }
 
+    private boolean hasMoreMessages(MessageId lastMessageIdInBroker, MessageId lastDequeuedMessage) {
+        return (lastMessageIdInBroker.compareTo(lastDequeuedMessage) > 0 ||
+                // Make sure batch message can be read completely.
+                lastDequeuedMessage instanceof BatchMessageIdImpl
+                        && lastMessageIdInBroker.compareTo(lastDequeuedMessage) >= 0
+                        && incomingMessages.size() > 0)
+                && ((MessageIdImpl)lastMessageIdInBroker).getEntryId() != -1;
```

Comment:
I think the condition is not correct in current form. I believe it should be like : 

```java
if (lastMessageIdInBroker.compareTo(lastDequeuedMessage) > 0 
    ((MessageIdImpl)lastMessageIdInBroker).getEntryId() != -1) {
    return true;
} else if (lastMessageIdInBroker.compareTo(lastDequeuedMessage) == 0
               && incomingMessages.size() > 0 ) {
     return true;
} else {
   return false;
}
```

Reply:
@merlimat Thanks, more clearer.